base_layer/core/src/covenants/covenant.rs

```diff
@@ -39,42 +39,38 @@ use crate::{
     transactions::transaction_components::{TransactionInput, TransactionOutput},
 };
 
-const MAX_COVENANT_BYTES: usize = 4096;
-
 #[derive(Debug, Clone, PartialEq, Eq, Default)]
 pub struct Covenant {
     tokens: Vec<CovenantToken>,
 }
 
-impl Covenant {
-    pub fn new() -> Self {
-        Self { tokens: Vec::new() }
-    }
-
-    pub fn from_bytes(mut bytes: &[u8]) -> Result<Self, CovenantDecodeError> {
-        if bytes.is_empty() {
-            return Ok(Self::new());
+impl BorshSerialize for Covenant {
+    fn serialize<W: Write>(&self, writer: &mut W) -> io::Result<()> {
+        BorshSerialize::serialize(
+            &(u32::try_from(self.tokens.len()).map_err(|_| ErrorKind::InvalidInput)?).to_le_bytes(),
+            writer,
+        )?;
+        for token in self.tokens.iter() {
+            BorshSerialize::serialize(token, writer)?;
         }
-        if bytes.len() > MAX_COVENANT_BYTES {
-            return Err(CovenantDecodeError::ExceededMaxBytes);
-        }
-        CovenantTokenDecoder::new(&mut bytes).collect()
-    }
-
-    pub fn to_bytes(&self) -> Vec<u8> {
-        let mut buf = Vec::with_capacity(self.get_byte_length());
-        self.write_to(&mut buf).unwrap();
-        buf
+        Ok(())
     }
+}
 
-    pub(super) fn write_to<W: io::Write>(&self, writer: &mut W) -> Result<(), io::Error> {
-        CovenantTokenEncoder::new(self.tokens.as_slice()).write_to(writer)
+impl BorshDeserialize for Covenant {
+    fn deserialize(buf: &mut &[u8]) -> io::Result<Self> {
+        let len = u32::deserialize(buf)?;
+        let mut tokens = Vec::with_capacity(len as usize);
+        for _ in 0..len {
+            tokens.push(BorshDeserialize::deserialize(buf)?);
```

Comment:
For covenants and tari script, I would rather use the encoding provided.

The type `&[u8]` implements `std::io::Read` so nothing much needs to change.

```rust
 let len = u32::deserialize(buf)? as usize;
        if len == 0 {
            return Ok(Covenant::new());
        };
        // Check the length  - this may be maliciously misreported
        if len > MAX_COVENANT_BYTES {
            return Err(io::Error::new(
                io::ErrorKind::InvalidInput,
                "covenants: Length varint exceeded maximum",
            ));
        }
        // Ensure that no more than the maximum bytes can be read
        let mut limited = LimitedBytesReader::new(MAX_COVENANT_BYTES, buf);
        CovenantTokenDecoder::new(&mut limited)
            .collect::<Result<_, CovenantDecodeError>>()
            .map_err(|err| io::Error::new(io::ErrorKind::InvalidInput, err))
```

Reply:
The Covenant (I haven't checked the TariScript) uses the consensus encoding in some parts (e.g.CovenantArg). That should be replace by borsh, right? Just checking, because that will slightly change the outputs.